### ISSUE TITLE
Refactoring the code for intermediate and init process handling

### ISSUE THREAD

**Issue by utam0k** [containers/youki#295]
# Goal
These two processing parts are very difficult to write tests for, but they are the places where tests should be written. I would like to expand the tests while refactoring them.

related https://github.com/containers/youki/issues/279

**Comment by utam0k** [comment]
I'm looking for ideas for refactoring.

**Comment by Furisto** [comment]
- For stuff like sysctls that just write to files introduce a file system facade that is replaced by a mock when testing
```
fn sysctl(fs: Box<Filesystem>, kernel_params: &HashMap<String, String>) -> Result<()> {
    let sys = PathBuf::from("/proc/sys");
    for (kernel_param, value) in kernel_params {
        let path = sys.join(kernel_param.replace(".", "/"));
        log::debug!(
            "apply value {} to kernel parameter {}.",
            value,
            kernel_param
        );
        fs.write(path, value.as_bytes())
            .with_context(|| format!("failed to set sysctl {}={}", kernel_param, value))?;
    }

    Ok(())
}
```
- Likewise for things like readonly paths we can extend the syscall mock that we already have to e.g. check that the expected mounts have been executed with the right parameters. With proper use of generics we can even avoid having to box the abstractions.
- Another option could be to just cfg(test) the methods e.g. 
```
struct Filesystem {
  fn write(&self, ..) -> Result<()> {
   #cfg(test)
   write_test();

   #cfg(not(test))
   write_real();
   }
}
```
- The rest will have to be covered by the integration tests

**Comment by utam0k** [comment]
@Furisto Thank you very much. These are issues that we need to think about too. I think I should also rethink the directory structure a bit. It would be nice to have a great name for the init and intermediate processes together. What do you think?

**Comment by Furisto** [comment]
The structure seems alright to me, how would you like to change it?

**Comment by utam0k** [comment]
> The structure seems alright to me, how would you like to change it?

I think that at least init and intermediate should be separated by file or module.

**Comment by chenyukang** [comment]
I think the structure is alright, but why cgroup is in top level of the repo?

**Comment by utam0k** [comment]
I'm going to try not to make too many changes since people didn't seem to find it a big problem. Thanks for the input.

This is because we believe that group can eventually be provided as a crate, just like oci-spec-rs. Also, having a separate crate has the advantage of speeding up compilation in some case.
> I think the structure is alright, but why cgroup is in top level of the repo?

**Comment by yihuaf** [comment]
First of all, I think this code path is heavily tested by the integration tests, so the priority is not that high in terms of confidence that the code works. With that being said, we should definitely increase unit tests, even just for better debugging for ourselves. Mocking syscall is one way, but I think it will only making sure the syscall is invoked. To be more thorough, we need to
1. fork a process and perhaps entering into namespaces depending on the test case.
2. parent process wait result from the child process.
3. invoke the syscall or operation we want to test in the child. Take `readonly_path` for example.
4. Actually test the path is readonly by writing to it.
5. send result from child to parent. parent should clean up and return the test results.

Another way to do this is have integration test somehow generate the .gcov files. I have only done it with `C` code but I believe rust coverage works in similar ways. I don't have all the details, but I believe how coverage works is by compiling binary with a special flag, which will generate .gcov files when binary gets executed. The coverage then looks at all the .gcov files and decide how much code was executed. We can compile Youki with the coverage flag and run integration tests. Maybe that's a good enough substitute for unit testing?

**Comment by utam0k** [comment]
> Another way to do this is have integration test somehow generate the .gcov files. I have only done it with `C` code but I believe rust coverage works in similar ways. I don't have all the details, but I believe how coverage works is by compiling binary with a special flag, which will generate .gcov files when binary gets executed. The coverage then looks at all the .gcov files and decide how much code was executed. We can compile Youki with the coverage flag and run integration tests. Maybe that's a good enough substitute for unit testing?

Wow, I'm interested in this. However, we can get feedback faster by running unit tests than by running integration tests. I'm not saying perfect code coverage, but I think we should try to do as much unit test effort as possible. I believe this will help future developers include me and encourage loosely coupled code.

**Comment by yihuaf** [comment]
> > Another way to do this is have integration test somehow generate the .gcov files. I have only done it with `C` code but I believe rust coverage works in similar ways. I don't have all the details, but I believe how coverage works is by compiling binary with a special flag, which will generate .gcov files when binary gets executed. The coverage then looks at all the .gcov files and decide how much code was executed. We can compile Youki with the coverage flag and run integration tests. Maybe that's a good enough substitute for unit testing?
> 
> Wow, I'm interested in this. However, we can get feedback faster by running unit tests than by running integration tests. I'm not saying perfect code coverage, but I think we should try to do as much unit test effort as possible. I believe this will help future developers include me and encourage loosely coupled code.

I understand. And I agree we should write as many unit tests as possible, and that will allow us to think about the test-ability of our code. In that way, I think we should be mindful about breaking syscalls into their own functions and we can test that. Example would be `readonly_path` and etc. For the intermediate and init process, I think `mock` syscalls is probably easier to do.

**Comment by utam0k** [comment]
That's right. My initial codes were not well organized in this area, and I ended up in this state. I'd like to organize them out a bit. By mock syscalls, do you mean the current `src/syscall/tests.rs`?
> In that way, I think we should be mindful about breaking syscalls into their own functions and we can test that. Example would be readonly_path and etc. For the intermediate and init process, I think mock syscalls is probably easier to do.

**Comment by yihuaf** [comment]
>  By mock syscalls, do you mean the current src/syscall/tests.rs?

That's what I imagined. If we use test syscall to mock everything, then we would run the intermediate and init to completion without invoking the actual syscall. At the end, we just check if each syscall is called. Not sure how much this will achieve though, compared to the integration test which will test for the correctness. Breaking the syscall out into their own testable unit probably will achieve more in terms of increasing our confidence in the code correctness.

**Comment by utam0k** [comment]
Yes, I totally agree with your opinion. In the long run, I believe unit tests will benefit us differently than integration tests.
